include/seahorn/Transforms/Instrumentation/ShadowMemSeaDsa.hh

```diff
@@ -1,22 +1,51 @@
 /// This file should be completely moved into C++ file.
+// TODO: remove this comment now? we want to use this pass in the vcgen
 #pragma once
 #include "llvm/Pass.h"
 
+#include "sea_dsa/Global.hh"
+#include "sea_dsa/Graph.hh"
+
+#include <memory>
+
 namespace llvm {
 class Value;
 }
+
+namespace {
+// defined in ShadowMemSeaDsa.cc
+class ShadowDsaImpl;
+}
+
 #define USE_NEW_SHADOW_SEA_DSA 1
 namespace seahorn {
+
 class ShadowMemSeaDsa : public llvm::ModulePass {
+
+private:
+  sea_dsa::GlobalAnalysis *m_dsa;
+  std::unique_ptr<ShadowDsaImpl> m_shadow;
+
 public:
   static char ID;
-  ShadowMemSeaDsa() : llvm::ModulePass(ID) {}
+  ShadowMemSeaDsa();
+  ~ShadowMemSeaDsa();
 
   bool runOnModule(llvm::Module &M) override;
 
   bool runOnFunction(llvm::Function &F);
   void getAnalysisUsage(llvm::AnalysisUsage &AU) const override;
   llvm::StringRef getPassName() const override { return "ShadowMemSeaDsa"; }
+
+  // TODO: should this be a cell?
+  bool hasShadowId(const sea_dsa::Node *);
```

Comment:
I am not sure you can/need to go from DSA nodes to shadow ids. But you do want the opposite, 
given a `shadow.mem` call site, you want to know what DSA nodes it corresponds to. I would do this 
by providing a class, like CallSite does, that wraps a shdow.mem instruction and provides the necessary information to clients.

Reply:
I think that this is needed for efficiency. When traversing the BU graph, it's useful to get directly the id of all the reachable nodes from each actual parameter in the callsite, so that we know where to copy.

Reply:
I need to understand the use case better. Note that there is no one-to-one correspondence between shadow ids and nodes. Different fields of a node might have different ids. Of course we can ignore this in the first implementation, but it would be nicer if we have an interface that makes it simpler.

Reply:
Yes, in fact, I think this should be in terms of cells, as you mention, different fields may have different ids.
I was considering first the case in which only one logical array is created per dsa node (that is why I wrote the comment) because I thought that it was easier, but I will change it to cell.